File: common-core/src/main/java/io/quartic/common/uid/UidUtils.java
package io.quartic.common.uid;

import java.util.Collection;
import java.util.List;

import static java.util.stream.Collectors.toList;

public final class UidUtils {
    private UidUtils() {}

    public static String stringify(Uid uid) {
        return uid.uid();
    }

    public static List<String> stringify(Collection<? extends Uid> uids) {
        return uids.stream().map(UidUtils::stringify).collect(toList());
    }
}

File: common-core/src/main/kotlin/io/quartic/common/uid/UidUtils.kt
package io.quartic.common.uid

import org.apache.commons.codec.binary.Hex
import java.security.SecureRandom
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

fun <T : Uid> sequenceGenerator(converter: (String) -> T) = object : UidGenerator<T> {
    val counter = AtomicInteger()
    override fun get() = converter(counter.incrementAndGet().toString())
}

fun <T : Uid> randomGenerator(converter: (String) -> T) = object : UidGenerator<T> {
    override fun get() = converter(UUID.randomUUID().toString().substring(0, 6))
}

private val random = SecureRandom()
fun <T : Uid> secureRandomGenerator(converter: (String) -> T) = object : UidGenerator<T> {
    override fun get(): T {
        val bytes = ByteArray(16)    // Length-128
        random.nextBytes(bytes)
        return converter(Hex.encodeHexString(bytes))
    }
}

